new security guidelines & info (#54)
swift-server/sswg: security.md

# Security

> Security is at the heart of a healthy and dependable software ecosystem, and therefore also at the heart of the SSWG. The documents linked from this page intend to provide clear and simple guidance for both users and developers of packages listed in the SSWG's package index.

Reviewer: Normally I'd advise _away_ from passive voice, but in this case it sounds better to me 🤷‍♀️ 
```suggestion
Security is at the heart of a healthy and dependable software ecosystem, and therefore also at the heart of the SSWG. The documents linked from this page are intended to provide clear and simple guidance for both users and developers of packages listed in the SSWG's package index.
```

## Available information

> The information we have is split into different scenarios, depending on why you are here. Each piece of information is supposed to be self-contained, so please select the most appropriate one. If the reason you are here is not covered, please get in touch with the [SSWG on the forums](https://forums.swift.org/c/server/serverdev/14). If you're just curious, feel free to read it all at once 😊.

Reviewer 2: ```suggestion
The information we have is split into different scenarios, depending on the use case. Each piece of information is supposed to be self-contained, so please select the most appropriate one. If your use case is not covered, please get in touch with the [SSWG on the forums](https://forums.swift.org/c/server/serverdev/14). If you're just curious, feel free to read it all at once 😊.
```

Author: Thanks, wasn't a big fan of "use case" here, so reworded it to

```markdown
The information is split into different scenarios, depending on what brings you here. Each piece of information is supposed to be self-contained, so please select the most appropriate one. If your situation isn't covered, please get in touch with the [SSWG on the forums](https://forums.swift.org/c/server/serverdev/14).
```

Author: does that work for you?

> - If you found (or suspect you did) a security vulnerability in any of the packages listed on the SSWG's [package index](https://swift.org/server/#projects), have a look at our information regarding [reporting a security vulnerability](security/contributor-found-vulnerability.md).

Reviewer 2: ```suggestion
- If you found (or suspect you did) a security vulnerability in any of the packages incubated by the SSWG and listed on the SSWG [package index](https://swift.org/server/#projects), have a look at the information regarding [reporting a security vulnerability](security/contributor-found-vulnerability.md).
```

- If you are a package maintainer and received a report of a (suspected) security vulnerability, check out our guidance around [receiving notice of a security vulnerability](security/package-maintainer-received-vulnerability-report.md).

> - If you are a package maintainer of any of the listed projects or are thinking about pitching/proposing for your package to be listed, please find information about the [package security requirements](security/package-requirements.md).

Reviewer 2: ```suggestion
- If you are a package maintainer of any of the incubated projects or are thinking about pitching/proposing for your package to be listed, please find information about the [package security requirements](security/package-requirements.md).
```

# Security

> Security is at the heart of a healthy and dependable software ecosystem and therefore also at the heart of the SSWG. This page intends to give clear and easy to follow information to users and developers of packages listed on the SSWG's package index.

Reviewer: Grammar/phrasing nits:
```suggestion
Security is at the heart of a healthy and dependable software ecosystem, and therefore also at the heart of the SSWG. This document intends to provide clear and simple guidance for both users and developers of packages listed in the SSWG's package index.
```

> If you find any information to be missing, unclear, or misleading, please contact any member of the SSWG or create a post in the [server area](https://forums.swift.org/c/server/serverdev/14) of the Swift Forums.

Reviewer: Grammar/phrasing nits:
```suggestion
To report any missing, unclear, or misleading information, or any other concerns, please contact any member of the SSWG. You can also create a post in the [server area](https://forums.swift.org/c/server/serverdev/14) of the Swift Forums.
```

## I found a security vulnerability, what should I do?

> If you found a security vulnerability, first of all thank you for caring about our ecosystem. We rely on the community to keeping us all as safe and secure as we can (at least when it comes to software).

Reviewer 3: I think this should be a bit broader to include suspected vulnerabilities too.

The community bit of this paragraph is quite similar to the very first paragraph, in this document I'm not sure it's necessary here.

Author: @glbrntt +1 on suspected vulns. The repetition is deliberate because this page addresses different personas such as "I am a random person and found a security vuln, what should I do" as well as "I am a maintainer, what do I need to do when a vuln report happened". I think that's not clear that nobody is expected to read the whole doc but rather only the section that matches where they are right now.

Author: Regarding the repetition: Maybe we should have a security landing page + links like "I found a vulnerability, what should I do?" / "SSWG Security requirements for package authors" / "I am a package author, somebody reported a vuln, what's next"

Reviewer 3: Yeah, landing page and separate files or just making the sections more obvious with dividers and a contents would also work.

Author: done

> If you found a security vulnerability, first of all thank you for caring about our ecosystem. We rely on the community to keeping us all as safe and secure as we can (at least when it comes to software).

Reviewer: Grammar/phrasing nits:
```suggestion
If you've found a security vulnerability, first of all thank you for caring about our ecosystem! We rely on the community's participation to keep us all as safe and secure as possible, at least when it comes to software.
```

> The SSWG only takes a secondary role in the security process. The SSWG makes sure that listed projects do have a security process and that they follow their own process but the SSWG is typically not the first point of contact. Please don't read this as sending you away though, if at any point in the process you feel like you don't know where to report, you don't get a response, or face any other issues, please do contact the SSWG at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org) (this can only be seen by the SSWG members and the Swift core team).

Reviewer: Grammar/phrasing nits:
```suggestion
The SSWG takes only a secondary role in the security process. We make sure that listed projects have a security process, and that they follow their processes, but the SSWG is typically not the first point of contact. Please don't read this as sending you away, though; if at any point in the process you feel like you don't know where to report, you don't get a response, or have any other issues, please do contact the SSWG at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org). Emails sent to this address are never public; only SSWG members and the Swift core team can see them.
```

> 

Reviewer 3: I think it may be better to switch these paragraphs so you point reporters to individual packages first, then the fallback is communicating with the SSWG directly. I also think this section would benefit from being much more concise.

Author: thanks, done. Better?

> From packages on the SSWG's package index, it is required to list a clear point of contact on the project's main information page. Typically that is the `README.md` file in the project's GitHub/GitLab/... repository. It is also required that the project list the exact steps on how to report a vulnerability. Please follow these steps, the SSWG will be notified by the project directly. If you can't find any required information, you don't get a response from the project's maintainers, or you a are unsure that the vulnerability is treated with the appropriate care and urgency, please do contact the SSWG at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org).

Reviewer 3: Why README.md and not SECURITY.md?

Reviewer 3: `or you a are unsure` -> `or you are unsure`

Author: > Why README.md and not SECURITY.md?

I want it to be as easy as possible to find. If people link `SECURITY.md` from `README.md` that sounds great too. I think Github can also help you to find `SECURITY.md`, right? If yes, I SECURITY.md + github help may be even better.

Reviewer 4: May be worth noting that there _may_ be a `SECURITY.md` file and it's worth looking for it.

> From packages on the SSWG's package index, it is required to list a clear point of contact on the project's main information page. Typically that is the `README.md` file in the project's GitHub/GitLab/... repository. It is also required that the project list the exact steps on how to report a vulnerability. Please follow these steps, the SSWG will be notified by the project directly. If you can't find any required information, you don't get a response from the project's maintainers, or you a are unsure that the vulnerability is treated with the appropriate care and urgency, please do contact the SSWG at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org).

Reviewer: Grammar/phrasing nits:
```suggestion
Packages on the SSWG's package index are required to list a clear point of contact on the project's main information page. Most often, this will be the `README.md` file at the root of the project's GitHub/GitLab/etc. repository. Projects are also required to list exact steps for reporting a vulnerability. Please follow these steps; the SSWG will be notified by the project directly. If you can't find any required information, you don't get a response from the project's maintainers, or you feel that the vulnerability is not being treated with the appropriate care and urgency, please reach out to the SSWG at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org).
```

## For package maintainers: SSWG Security requirements

Packages that are listed on the [SSWG's package index](https://swift.org/server/#projects) are required to follow the following guidelines around security.

> When it comes to security vulnerabilities it is key to make sure that somebody who discovers a vulnerability in your package can learn how to report very quickly. As the package author, you know best where to put important information around your software. In general it is probably the `README.md` file in the root of your repository on GitHub, GitLab, or similar. Below "main information page" will refer to the `README.md` or a similar easy-to-find resource with information about your project.

Reviewer 3: I'd suggest calling out `SECURITY.md` as well since GitHub special cases it so that it is prominent when creating issues and so on (https://docs.github.com/en/github/managing-security-vulnerabilities/adding-a-security-policy-to-your-repository#about-security-policies).

Author: Ok, sounds great!

> When it comes to security vulnerabilities it is key to make sure that somebody who discovers a vulnerability in your package can learn how to report very quickly. As the package author, you know best where to put important information around your software. In general it is probably the `README.md` file in the root of your repository on GitHub, GitLab, or similar. Below "main information page" will refer to the `README.md` or a similar easy-to-find resource with information about your project.

Reviewer: Grammar/phrasing nits:
```suggestion
Where security vulnerabilities are involved, it is key to ensure that somebody who discovers a vulnerability in your package can immediately find information on how to report it. As the package author, you know best where to put important information about your software. In general, it is typically the `README.md` file in the root of your repository on GitHub/GitLab/etc. The phrase "main information page" is used elsewhere in this document to refer to the `README.md` or other resource as appropriate for your project.
```

The key requirements are:

> - Have a clear point of contact (eg. email address) for security vulnerability reporting listed on your main information page.
> - The point of contact for security vulnerabilities should be private. That means it is only accessible to your maintainers and _not_ publicly accessible on the internet. It is also good practise to describe exactly who can access the information next to the point of contact.
> - Either on the main information page or a page linked from it, describe as easily and precisely as possible how the security process of your package works. This includes where and how to report as well as expected timelines for a fix.
> - The goal is to make it as easy as possible for a user to report a security vulnerability to you. Try to remove all ambiguity and difficulty from reporting vulnerabilities. In many cases you will rely on somebody's spare time to report a vulnerability to you.

Reviewer: Grammar/phrasing nits:
```suggestion
- List a clear point of contact (e.g. an email address) for security vulnerability reports on your main information page.
- The point of contact for security vulnerabilities should be private. That means it is only accessible to you and your maintainers, and in particular that it is _not_ publicly accessible on the Internet. It is also good practice to describe exactly who can access the information alongside the point of contact.
- Either on the main information page or a page linked from it, describe as easily and precisely as possible how the security process of your package works. This includes where and how to report as well as expected timelines for a fix.
- The goal is to make it as easy as possible for a user to report a security vulnerability to you. Try to remove all ambiguity and difficulty from reporting vulnerabilities. In many cases you will rely on somebody's spare time to report vulnerabilities.
```

## For Package maintainers: What to do if there is a security vulnerability

> The discovery of a security vulnerability in your code can be daunting but it is part of the life of any software developer. So first of all, don't stress. If you need any help at any step of the way, please feel free to contact any of the SSWG members or the group as a whole at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org).

Reviewer: Grammar/phrasing nits:
```suggestion
The discovery of a security vulnerability in your code can be daunting, but it is part of the life of any software developer. So first of all, don't stress. If you need any help at any step of the way, please feel free to contact any of the SSWG members or the group as a whole at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org).
```

> Here is a step-by-step guide on what to do with regards to the SSWG.

Reviewer: ```suggestion
Here is a step-by-step guide on what to do with regards to the SSWG:
```

> 1. The SSWG is only a secondary point of contact which can broadcast information about vulnerabilities. The best way to start is to start addressing the vulnerability according the the security process of your own package.

Reviewer 3: I'm not sure what the entire role of the SSWG is here, it seems to be a few different things: keeping "in the loop", providing guidelines for package authors/maintainers and acting as a fallback for when packages don't have enough information on how to deal with vulnerabilities. Is there anything else?

re: the first part, what does the SSWG do with the information? Previously there was a note about packages failing to appropriately deal with issues having their status lowered or removed from the package index. Is that intentionally dropped?

Author: > I'm not sure what the entire role of the SSWG is here, it seems to be a few different things: keeping "in the loop", providing guidelines for package authors/maintainers and acting as a fallback for when packages don't have enough information on how to deal with vulnerabilities. Is there anything else?
> 
> re: the first part, what does the SSWG do with the information? Previously there was a note about packages failing to appropriately deal with issues having their status lowered or removed from the package index. Is that intentionally dropped?

I think that's pretty much it.

> re: the first part, what does the SSWG do with the information?

- verify that processes work and are followed
- being aware that something potentially dangerous is out there, so in certain cases we may want to commit resources to make an extra effort to contact known adopters

>  Previously there was a note about packages failing to appropriately deal with issues having their status lowered or removed from the package index. Is that intentionally dropped?

I thought it sounded like a threat and maybe a bit over the top. I thought that if a package doesn't follow the guidelines it's kinda obvious that it may be demoted and doesn't need to be called out here. But maybe it's good to explicitly state it.

Reviewer 3: Cool, I think it might be worth explicitly stating the role of the SSWG in the section for maintainers. If anything it's useful to make it more obvious that they are there to offer guidance and help when necessary.

---

> I thought it sounded like a threat and maybe a bit over the top. I thought that if a package doesn't follow the guidelines it's kinda obvious that it may be demoted and doesn't need to be called out here. But maybe it's good to explicitly state it.

It _is_ a threat though isn't it? No strong feelings whether it's over the top or not.

Reviewer 5: I think it isn't necessary in the document, that's something better kept outlined in the Incubation Process. I feel this doc should be entirely focused on what @weissi said: A place to point to for different personas on how to move forward with vulnerabilities

> 2. As soon as you can -- but within 10 days of discovering / receiving notification about the vulnerability -- please notify the SSWG about the vulnerability at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org). The SSWG will not discolose any information about this vulnerability and the emails can only be seen by the SSWG members listed on [swift.org/server](https://swift.org/server/) as well as the [Swift Core team](https://swift.org/community/#community-structure). If you prefer to share the vulnerability only with a smaller group, please feel free to reach out to any of the SSWG members individually.

Reviewer 4: 10 business days or 10 calendar days? We should be unambiguous here.

> 2. As soon as you can -- but within 10 days of discovering / receiving notification about the vulnerability -- please notify the SSWG about the vulnerability at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org). The SSWG will not discolose any information about this vulnerability and the emails can only be seen by the SSWG members listed on [swift.org/server](https://swift.org/server/) as well as the [Swift Core team](https://swift.org/community/#community-structure). If you prefer to share the vulnerability only with a smaller group, please feel free to reach out to any of the SSWG members individually.

Reviewer 4: ```suggestion
2. As soon as you can -- but within 10 days of discovering / receiving notification about the vulnerability -- please notify the SSWG about the vulnerability at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org). The SSWG will not disclose any information about this vulnerability and the emails can only be seen by the SSWG members listed on [swift.org/server](https://swift.org/server/) as well as the [Swift Core team](https://swift.org/community/#community-structure). If you prefer to share the vulnerability only with a smaller group, please feel free to reach out to any of the SSWG members individually.
```

> 3. After fixing the vulnerability, please promptly (within three days of fixing / publishing the vulnerability) create a new Swift Forums post in the [Server > Security Updates](https://forums.swift.org/c/server/security-updates/50) category, linking to your own security advisory. The security advisory should contain at least which versions of what software are affected and how to update to an unaffected version.

Reviewer 3: Fixing or releasing a patched version?

Author: release

> From [_graduated_](https://github.com/swift-server/sswg/blob/master/process/incubation.md#process-diagram) projects The SSWG typically expects that the whole process -- from the initial report/discovery to fixing and publishing the vulnerability -- is completed within 30 days. We do however acknowledge that certain types of vulnerabilities are either very complicated to address or a part of a "coordinated disclosure" which means that 30 days may not be enough. That is absolutely understandable, please however make sure to inform the SSWG (at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org)) about your anticipated timelines and any significant divergence from the plan.

Reviewer 4: ```suggestion
From [_graduated_](https://github.com/swift-server/sswg/blob/master/process/incubation.md#process-diagram) projects the SSWG typically expects that the whole process -- from the initial report/discovery to fixing and publishing the vulnerability -- is completed within 30 days. We do however acknowledge that certain types of vulnerabilities are either very complicated to address or a part of a "coordinated disclosure" which means that 30 days may not be enough. That is absolutely understandable, please however make sure to inform the SSWG (at [sswg-security-reports@forums.swift.org](mailto:sswg-security-reports@forums.swift.org)) about your anticipated timelines and any significant divergence from the plan.
```

> Failure to report or address vulnerabilities may result in the SSWG publishing a security advisory, and could lead to retracting the project's status and listing it under the non-recommended projects list. In some cases, the SSWG may choose to find a technical contributor that can help resolve the security issues to minimize the impact. SSWG actions will be decided on a case by case basis and require a supermajority vote.

Reviewer 4: Clarify that this means "failure to report vulnerabilities _to the community_ or _to the SSWG_", not a failure of a user to report a vulnerability to a project.
